lib/models/plugin-registry.js

```diff
@@ -0,0 +1,205 @@
+    var installedPlugins  = this._installedPlugins(validAddons, aliasMap);
+    var enabledPlugins    = this._applyDisabledConfig(installedPlugins, disabledMap);
+    var orderedPlugins    = this._applyRunOrderConfig(enabledPlugins, runOrderMap);
+
+    if (orderedPlugins.length === 0) {
+      this._ui.writeError('\nWARNING: No plugins installed/enabled\n');
+      this._ui.writeError('ember-cli-deploy works by registering plugins in it\'s pipeline.\n');
+      this._ui.writeError('In order to execute a deployment you must install at least one ember-cli-deploy compatible plugin.\n');
+      this._ui.writeError('Visit http://ember-cli-deploy.github.io/ember-cli-deploy/docs/v0.6.x/plugins/ for a list of supported plugins.\n');
+    }
+
+    return orderedPlugins;
+  },
+
+  _validAddons: function(addons) {
+    if (!this._cachedValidAddons) {
+      this._cachedValidAddons = this._discoverValidAddons(addons);
+    }
+
+    return this._cachedValidAddons;
+  },
+
+  _discoverValidAddons: function(addons){
+    var self = this;
+    return addons.reduce(function(validAddons, currentAddon) {
+      var name = self._pluginNameFromAddonName(currentAddon);
+
+      if (self._isValidDeployPlugin(currentAddon)) {
+        validAddons[name] = currentAddon;
+      }
+
+      if (self._isDeployPluginPack(currentAddon)) {
+        var nestedAddons = self._discoverValidAddons(currentAddon.addons);
+        validAddons = Object.keys(nestedAddons).reduce(function(v, key) {
+          v[key] = nestedAddons[key];
+          return v;
+        }, validAddons);
+      }
+
+      return validAddons;
+    }, {});
+  },
+
+  _buildAliasMap: function(addons, config) {
+    var self = this;
+    var aliasMap = Object.keys(addons)
+      .reduce(function(aliases, name) {
+        if (!aliases[name] || !aliases[name].as) {
+          aliases[name] = { as: [name] };
+        }
+
+        aliases[name].as = self._asArray(aliases[name], 'as');
+        return aliases;
+      }, config);
+
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(aliasMap), addons, 'config.pipeline.alias');
+
+    return aliasMap;
+  },
+
+  _buildDisabledMap: function(aliasMap, config, addons) {
+    var aliasArray = Object.keys(aliasMap)
+      .map(function(addonName) {
+        return aliasMap[addonName].as;
+      })
+      .reduce(function(aliases, arr) {
+        return aliases.concat(arr);
+      }, []);
+
+    var disabledMap = aliasArray.reduce(function(map, alias) {
+      if (map[alias] === undefined) {
+        map[alias] = false;
+      }
+
+      return map;
+    }, config);
+
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(disabledMap), addons, 'config.pipeline.disabled');
+
+    return disabledMap;
+  },
+
+  _buildRunOrderMap: function(config, addons) {
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(config), addons, 'config.pipeline.runOrder');
+
+    return config;
```

Comment:
That was my thinking ;)
On Tue, 28 Jun 2016 at 14:59, Mattia Gheda <notifications@github.com> wrote:

> In lib/models/plugin-registry.js
> <https://github.com/ember-cli-deploy/ember-cli-deploy/pull/407#discussion_r68761006>
> :
>
> > +      if (map[alias] === undefined) {
> > +        map[alias] = false;
> > +      }
> > +
> > +      return map;
> > +    }, config);
> > +
> > +    this._warnIfConfigReferencesUninstalledAddons(Object.keys(disabledMap), addons, 'config.pipeline.disabled');
> > +
> > +    return disabledMap;
> > +  },
> > +
> > +  _buildRunOrderMap: function(config, addons) {
> > +    this._warnIfConfigReferencesUninstalledAddons(Object.keys(config), addons, 'config.pipeline.runOrder');
> > +
> > +    return config;
>
> ok. this is what I thought
> I saw the steps lines but wanted to confirm..
> I guess we can leave it in case we ever need to do something w/it later
> and it's still a good place for the warning
> 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ember-cli-deploy/ember-cli-deploy/pull/407/files/a17adad46375d992db68a691e6a4f739e8c18201#r68761006>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAZb1F2P61RD7vIWWsC5aGd4YMKXbGbbks5qQSisgaJpZM4IvW7Y>
> .
>